### ISSUE TITLE
certain kinds of swear words / oaths should be blocked in guilds

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#8813]
~This is on hold until the PR for https://github.com/HabitRPG/habitica/issues/8062 is completed (automatic muting of slurs), since there'll probably be code in that PR that can be repurposed for this issue.~ EDIT: We're now considering which words to block where.

In https://github.com/HabitRPG/habitica/pull/8197 , we prevented swear words, slurs, etc being posted to the Tavern and that's been working very well. We want it extended to public guilds, although with a more limited set of words. For example, in the Tavern, it's not permitted to mention legal addictive substances such as certain drinks, however those words would not be blocked in public guilds. 

We'll give more details about what should be blocked elsewhere after https://github.com/HabitRPG/habitica/issues/8062  is fixed. For now, this is just a placeholder issue so that we don't forget about this.

**Comment by abatula** [comment]
I can work on this if you have more details on what you need done. As originally written, my fix (#8177) to #8062 should also block those words in guilds and private groups. Is this a case of more extensive tests needed (e.g. test both tavern chat and a regular guild to make sure the words are blocked)?

**Comment by Alys** [comment]
@abatula Thank you for offering! However we'll leave this on hold for a little longer. As you might have seen from a [Bailey message](https://habitica.com/static/old-news), a new design for the website will be released soon. One aspect of it will be tags for guilds, to categorise them by topic. @shanaqui had the idea to use those tags to control swearword blocking in guilds. For example, most guilds would not be allowed to use religious oaths, however a guild tagged as being about religion would be allowed to use them because it would be more likely that the words were not being used as oaths.

So after the new site is released, moderators will group the current swearwords into sections associated with guild tags and then we'll take this issue off hold. If you'd like to work on it then, you'd be welcome to! If we remember, we'll post here to ask you about it then.

**Comment by Alys** [comment]
Labels updated.

@abatula I've marked this as in progress in case you still want to work on it, but if you don't no worries at all - just let us know and we'll put it back to help wanted.

We're actually doing a quick fix to introduce swearword blocking to most guilds, requested by Lemoness: https://github.com/HabitRPG/habitica/pull/9253 however that's intended to be just a short term solution until a better fix is done for this issue. We still like shanaqui's idea from my previous comment here. What I'm thinking is we'll have a **new** file to replace `website/server/libs/bannedWords.js` and it will have different arrays in it, something like this:

```
let bannedWordsReligious = [
  'TEST_PLACEHOLDER_BANNED_WORD_RELIGIOUS',
];
let bannedWordsSwears = [
  'TEST_PLACEHOLDER_BANNED_WORD_SWEARWORD',
];
let bannedWordsAddictive = [
  'TEST_PLACEHOLDER_BANNED_WORD_ADDICTIVE',
];
etc
```
Although if you had something else in mind, don't feel that you have to use that structure. 

You would attach the new file to your PR with nothing but test words in it, and then after your PR was merged, admins would do a direct commit to put the real words into the file and to delete the old `website/server/libs/bannedWords.js` file. This approach means that the banned words themselves wouldn't be visible in your PR, to protect you from seeing them all the time and to protect other contributors who might be looking through GitHub.

We still need to match up the guild categories with the lists of banned words, but as an example to start with, in any guild categorised under the "spirituality" category, you could post words from the `bannedWordsReligious` array, but not from any of the other arrays. 

It's expected that some guilds would have two or more categories that matched to two or more lists of banned words. For example, a guild focused on prayer to assist with addiction recovery would have appropriate categories to allow you to post words from bannedWordsReligious and bannedWordsAddictive but not from other lists.

Let us know what you think about this, or if you have any questions or other ideas!

**Comment by abatula** [comment]
That sounds like an interesting method, I'll take a look at it! Sorry for the slow response, I need to convince GitHub to email my other address.

**Comment by abatula** [comment]
Two questions:

Should I add a new file for banned words with a new name? If I replace the existing one, won't it show the current list of blocked words in the summary of changes (which you're trying to avoid)?

Also, should guilds be blacklisting the sets of words that aren't allowed, or white listing the ones that are? My first thought is to use blacklisting, but I think either way would work about the same.

**Comment by Alys** [comment]
@abatula Yes, please use a new file.

For blacklisting versus whitelisting:
We'll start with a few different lists of banned words, but we might add more lists in future. When we do, we'll want the new lists to be automatically applied to all existing guilds. Because of that, every list should apply to every guild unless that guild has a setting to say that a particular list does not apply. 
So I guess that means we'll need to whitelist a specific list when that list doesn't apply to the guild.

**Comment by Alys** [comment]
Just to ensure we don't forget about this, the PR that fixes this issue should delete the `guildsAllowingBannedWords` variable and any code that uses it. It was added as a short-term fix in https://github.com/HabitRPG/habitica/pull/9253 which has just now been merged into `develop`. FYI @abatula

**Comment by beffymaroo** [comment]
@Alys @abatula Is this issue still in process?

**Comment by abatula** [comment]
I haven't had a chance to fix my local install, so I haven't been working on it.

**Comment by beffymaroo** [comment]
@abatula OK. I'm adding a "help wanted" label to this issue, but if you find time and would like to pick it up again just let us know!

**Comment by TheHollidayInn** [comment]
This has been added, right?

**Comment by Alys** [comment]
A quick fix was done (mentioned in the comments) but we still need a proper solution. It's not high priority.

**Comment by Alys** [comment]
_I'm marking this issue as suggestion-discussion for the staff and mods to consider my proposed change below (in bold)._

The current state is that swear words are blocked in the Tavern and most public guilds. The only guilds that don't have swear word blocking are ones about special topics where normally blocked words do form part of legitimate conversations ([website/server/libs/guildsAllowingBannedWords.js](https://github.com/HabitRPG/habitica/blob/develop/website/server/libs/guildsAllowingBannedWords.js)). It's actually a pretty big list and it's not ideal to have it hard-coded like that, and in fact I believe I promised paglias that we wouldn't leave it like that for too long - sorry!! :scream: 

We had talked in the past about linking the handling of banned words with the categories that each guild is assigned to (e.g., alcohol would be allowed in cooking guilds) however moderators are now very aware that that wouldn't work because guild leaders often don't categorise their guilds correctly.

The mod team is actually pretty happy with how swear word blocking is working now so we don't really want changes to behaviour, but that long list of hard-coded guilds has to go. **I'm thinking it would be best if:**
* **Each public guild edit form had a checkbox for turning off swear word blocking (with an associated parameter in the group schema).**
* **The checkbox would be visible to only mods and staff (who already have the ability to edit any public guild).**
* **The code that decides whether to block swears would use that checkbox instead of the hard-coded list.**
* **Swear blocking would be on by default (as it is now).**
* **The list of hard-coded guilds would be deleted.**

If this proposal is acceptable, I'll close this issue and open a new one with a bit more detail in it (e.g., it'll need to be a two-step change with the modal support being done first so that we can mark appropriate guilds before the current mechanism is turned off).
